### PR TITLE
reduce our usage of guava collect to use jdk collections methods

### PR DESCRIPTION
**Pull request by olamy**
See [JENKINS-65053](https://issues.jenkins.io/browse/JENKINS-65053).

### Proposed changelog entries

* Reduce usage of guava collect api by using standard java 8 apis

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/model/LoadBalancer.java [hunk 1]
```diff
@@ -86,7 +86,7 @@ public Mapping map(Task task, MappingWorksheet ws) {
 
                 // Build a Map to pass in rather than repeatedly calling hash.add() because each call does lots of expensive work
                 List<ExecutorChunk> chunks = ws.works(i).applicableExecutorChunks();
-                Map<ExecutorChunk, Integer> toAdd = Maps.newHashMapWithExpectedSize(chunks.size());
+                Map<ExecutorChunk, Integer> toAdd = new HashMap( chunks.size());
```
[the remark this answers is not in the document]
Comment (basil) [reply core/src/main/java/hudson/model/LoadBalancer.java:89]: This creates an empty `HashMap` with the specified initial capacity and the default load factor (0.75). If you want to avoid rehashing the map, you'll need to do something like what Guava does internally:

```java
/**
 * Returns a capacity that is sufficient to keep the map from being resized as long as it grows no
 * larger than expectedSize and the load factor is ≥ its default (0.75).
 */
static int capacity(int expectedSize) {
  if (expectedSize < 3) {
    checkNonnegative(expectedSize, "expectedSize");
    return expectedSize + 1;
  }
  if (expectedSize < Ints.MAX_POWER_OF_TWO) {
    // This is the calculation used in JDK8 to resize when a putAll
    // happens; it seems to be the most conservative calculation we
    // can make.  0.75 is the default load factor.
    return (int) ((float) expectedSize / 0.75F + 1.0F);
  }
  return Integer.MAX_VALUE; // any large value
}
```

But really the use of `Maps.newHashMapWithExpectedSize` in the first place smells like an example of [premature optimization](https://en.wikiquote.org/wiki/Donald_Knuth). So my suggestion is to just use the default default initial capacity (16) and the default load factor (0.75). Best to write the simplest and clearest code and optimize only if and when necessary.

File: core/src/main/java/hudson/PluginWrapper.java [hunk 2]
```diff
@@ -220,7 +219,7 @@ public boolean hasDerivedDependencyErrors() {
      * The core can depend on a plugin if it is bundled. Sometimes it's the only thing that
      * depends on the plugin e.g. UI support library bundle plugin.
      */
-    private static Set<String> CORE_ONLY_DEPENDANT = ImmutableSet.copyOf(Collections.singletonList("jenkins-core"));
+    private static Set<String> CORE_ONLY_DEPENDANT = Collections.unmodifiableSet(new HashSet<>( Arrays.asList("jenkins-core")));
```
Comment (basil) [inline core/src/main/java/hudson/PluginWrapper.java:222]: Simpler:

```suggestion
    private static Set<String> CORE_ONLY_DEPENDANT = Collections.singleton("jenkins-core");
```

[Collections#singleton](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#singleton-T-) returns an immutable set.

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 3]
```diff
@@ -700,7 +700,7 @@ public final void post(BuildListener listener) throws Exception {
                 HashSet<String> r = new HashSet<>();
```
Comment (basil) [inline core/src/main/java/hudson/model/AbstractBuild.java:700]: Should we consider changing this to a sorted set to maintain the original behavior of sorting the culprits? Or is that not relevant in this use case? Not sure offhand.

```java
SortedSet<String> r = new TreeSet<>();
```

File: core/src/main/java/hudson/tasks/LogRotator.java [hunk 4]
```diff
@@ -217,7 +216,10 @@ public void perform(Job<?,?> job) throws IOException, InterruptedException {
             //Collate all encountered exceptions into a single exception and throw that
             String msg = String.format(
                     "Failed to rotate logs for [%s]",
-                    Joiner.on(", ").join(exceptionMap.keySet())
+                    exceptionMap.keySet().stream().
+                        filter(run -> run!=null).
+                        map(Run::toString).
+                        collect(Collectors.joining(", "))
```
Comment (basil) [inline core/src/main/java/hudson/tasks/LogRotator.java:222]: This code is fine. But perhaps a simpler alternative might be to use `hudson.Util`:

```java
Util.join(exceptionMap.keySet(), ", ")
```

I don't feel strongly one way or another.

File: test/src/test/java/jenkins/ClassPathTest.java [hunk 5]
```diff
@@ -51,7 +52,7 @@ public void uniqueness() throws Exception {
         for (File jar : new File(WarExploder.getExplodedDir(), "WEB-INF/lib").listFiles((dir, name) -> name.endsWith(".jar"))) {
             String jarname = jar.getName();
             try (JarFile jf = new JarFile(jar)) {
-                Iterators.forEnumeration(jf.entries()).forEachRemaining(e -> {
+                Iterators.forEnumeration( jf.entries()).forEachRemaining( e -> {
```
Comment (basil) [inline test/src/test/java/jenkins/ClassPathTest.java:55]: Looks like just whitespace changes here. Perhaps you meant something like this:

```java
for (JarEntry e : Collections.list(jf.entries())) {
```

File: core/src/main/java/hudson/util/RunList.java [hunk 6]
```diff
@@ -208,33 +216,9 @@ public R getLastBuild() {
...
-    @Deprecated
-    public RunList<R> filter(com.google.common.base.Predicate<R> predicate) {
-        size = null;
-        first = null;
-        base = Iterables.filter(base,predicate);
+        base = StreamSupport.stream(base.spliterator(), false).
+            filter(predicate).
+            collect( Collectors.toList());
         return this;
     }
```
Comment (basil) [inline core/src/main/java/hudson/util/RunList.java:223]: Unfortunately there is still a usage of this deprecated method; see jan-molak/jenkins-build-monitor-plugin#472. So before we can delete this deprecated method from Jenkins core, we need to update the baseline in Build Monitor View (to gain access to the non-deprecated replacement) and then land jan-molak/jenkins-build-monitor-plugin#472. In the meantime the deprecated method needs to remain in order to not break that plugin.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 7]
```diff
@@ -3198,7 +3196,7 @@ private void setBuildsAndWorkspacesDir() throws IOException, InvalidBuildsDir {
      * Checks the correctness of the newBuildsDirValue for use as {@link #buildsDir}.
      * @param newBuildsDirValue the candidate newBuildsDirValue for updating {@link #buildsDir}.
      */
-    @VisibleForTesting
+    //@VisibleForTesting
```
Comment (basil) [inline core/src/main/java/jenkins/model/Jenkins.java:3199]: This is in the `com.google.common.annotations` package, not the `com.google.common.collect` package, so it's technically out of the scope of this PR (which focuses on the `com.google.common.collect` package).

I don't feel strongly one way or another about removing it or keeping it. If we keep it, we could potentially use it with Checkstyle's [`VisibilityModifier`](https://checkstyle.sourceforge.io/config_design.html#VisibilityModifier) check in the future. I am working on a Checkstyle PR for this repository and I may look into this.

However, I do feel strongly that we remain consistent. This PR is inconsistent because it comments out `//@VisibleForTesting` in some cases but leaves it uncommented in other cases.

Overall, I think it would be best to revert the `@VisibleForTesting` changes in this PR and to open a separate PR to cover `com.google.common.annotations.VisibleForTesting` by either keeping it and taking advantage of it through static analyzers or else removing it consistently throughout the whole codebase.

File: test/src/test/java/hudson/security/TokenBasedRememberMeServices2Test.java [hunk 8]
```diff
@@ -128,7 +129,7 @@ public void basicFlow() throws Exception {
         wc.executeOnServer(() -> {
             Authentication a = Jenkins.getAuthentication2();
             assertEquals("bob", a.getName());
-            assertEquals(ImmutableList.of("authenticated", "myteam"), a.getAuthorities().stream().map(GrantedAuthority::getAuthority).collect(Collectors.toList()));
+            assertEquals(Arrays.asList("authenticated", "myteam"), a.getAuthorities().stream().map( GrantedAuthority::getAuthority).collect( Collectors.toList()));
```
Comment (basil) [inline test/src/test/java/hudson/security/TokenBasedRememberMeServices2Test.java:132]: [nit] Here and elsewhere, there seem to be some unrelated whitespace changes to code that isn't being changed by this PR. It would be ideal to revert such changes to eliminate noise and keep this PR focused on the important changes.

File: core/src/main/java/hudson/util/RunList.java [hunk 9]
```diff
@@ -24,9 +24,8 @@
 package hudson.util;
 
 import java.util.function.Predicate;
-import com.google.common.collect.Iterables;
-import com.google.common.collect.Iterators;
-import edu.umd.cs.findbugs.annotations.Nullable;
+
+import com.infradna.tool.bridge_method_injector.WithBridgeMethods;
```
Comment (basil) [inline core/src/main/java/hudson/util/RunList.java:28]: This import appears to be unused.

**Review by basil — COMMENTED** [review]
Great PR! This looks very good so far. I haven't yet reviewed the changes to `hudson.util.Iterators` and `hudson.util.RunList`, which are the most important parts of this PR, but I left some initial feedback on the rest of the PR to get you started.

File: core/src/main/java/hudson/model/LoadBalancer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (olamy) [reply core/src/main/java/hudson/model/LoadBalancer.java:89]: I don't have strong feeling here. using empty constructor (i.e default size/loadfactor) or using constructor with size will both rehash anyway :)

File: core/src/main/java/hudson/util/RunList.java (hunk 6 above)
Comment (olamy) [reply core/src/main/java/hudson/util/RunList.java:223]: doh not even a plugin hosted here...

File: test/src/test/java/jenkins/ClassPathTest.java (hunk 5 above)
Comment (olamy) [reply test/src/test/java/jenkins/ClassPathTest.java:55]: well this class has only one test and it;s marked as `@Ignored` I wanted to remove guava usage but then I gave up as it's not even executed. Anyway I will restore spaces...

File: core/src/main/java/hudson/model/AbstractBuild.java (hunk 3 above)
Comment (olamy) [reply core/src/main/java/hudson/model/AbstractBuild.java:700]: ah yes good point. not sure we really need it here but best to keep current behaviour

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 7 above)
Comment (olamy) [reply core/src/main/java/jenkins/model/Jenkins.java:3199]: sorry been to far enthusiast here 😄

File: core/src/main/java/hudson/util/RunList.java (hunk 6 above)
Comment (daniel-beck) [reply core/src/main/java/hudson/util/RunList.java:224]: Hosted "here" enough: https://plugins.jenkins.io/build-monitor-plugin/

File: core/src/main/java/jenkins/security/ClassFilterImpl.java [hunk 10]
```diff
@@ -106,7 +104,7 @@ private static void mockOff() {
         ClassFilter.setDefault(ClassFilter.NONE); // even Method on the standard blacklist is going to explode
     }
 
-    @VisibleForTesting
+    //@VisibleForTesting
```
Comment (basil) [inline core/src/main/java/jenkins/security/ClassFilterImpl.java:107]: To be consistent with the rest of this PR:

```suggestion
    @VisibleForTesting
```

File: core/src/main/java/hudson/util/Iterators.java [hunk 11]
```diff
@@ -410,4 +413,20 @@ public static void skip(@NonNull Iterator<?> iterator, int count) {
         }
     }
 
+    /**
+     *
+     * @param iterator the iterator
+     * @param position – position of the element to return
+     * @return the element at the specified position in iterator
+     */
+    public static <T> T get(Iterator<T> iterator, int position){
```
Comment (basil) [inline core/src/main/java/hudson/util/Iterators.java:422]: I'm curious why we're introducing this new API. Was there a particular need for this?

File: core/src/main/java/hudson/util/Iterators.java [hunk 12]
```diff
@@ -305,7 +305,8 @@ public void remove() {
      * Wraps another iterator and throws away nulls.
      */
     public static <T> Iterator<T> removeNull(final Iterator<T> itr) {
-        return com.google.common.collect.Iterators.filter(itr, Predicates.notNull());
+        Iterable<T> iterable = () -> itr;
+        return StreamSupport.stream(iterable.spliterator(), false).filter(t -> t != null).iterator();
```
Comment (basil) [inline core/src/main/java/hudson/util/Iterators.java:309]: Could use a method reference to `Objects::nonNull` rather than a custom lambda here:

```suggestion
        return StreamSupport.stream(iterable.spliterator(), false).filter(Objects::nonNull).iterator();
```

File: core/src/main/java/hudson/tasks/LogRotator.java [hunk 13]
```diff
@@ -217,7 +216,10 @@ public void perform(Job<?,?> job) throws IOException, InterruptedException {
             //Collate all encountered exceptions into a single exception and throw that
             String msg = String.format(
                     "Failed to rotate logs for [%s]",
-                    Joiner.on(", ").join(exceptionMap.keySet())
+                    exceptionMap.keySet().stream().
+                        filter(run -> run!=null).
```
Comment (basil) [inline core/src/main/java/hudson/tasks/LogRotator.java:220]: Could use a method reference to `Objects::nonNull` rather than a custom lambda here:

```suggestion
                        filter(Objects::nonNull).
```

File: core/src/main/java/hudson/util/Iterators.java [hunk 14]
```diff
@@ -316,7 +317,8 @@ public void remove() {
      */
     @SafeVarargs
     public static <T> Iterable<T> sequence( final Iterable<? extends T>... iterables ) {
-        return () -> new FlattenIterator<T,Iterable<? extends T>>(ImmutableList.copyOf(iterables)) {
+        return () -> new FlattenIterator<T,Iterable<? extends T>>(Collections.unmodifiableList(StreamSupport.stream(
+            Arrays.stream(iterables).spliterator(), true).collect( Collectors.toList()))) {
```
Comment (basil) [inline core/src/main/java/hudson/util/Iterators.java:321]: Simpler would be to use the `Stream` object returned by `Arrays.stream` rather than wrapping it in another stream:

```suggestion
        return () -> new FlattenIterator<T,Iterable<? extends T>>(Collections.unmodifiableList(
            Arrays.stream(iterables).collect(Collectors.toList()))) {
```

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 15]
```diff
@@ -697,10 +697,10 @@ public final void post(BuildListener listener) throws Exception {
                 post2(listener);
             } finally {
                 // update the culprit list
-                HashSet<String> r = new HashSet<>();
+                Set<String> r = new TreeSet<>();
```
[the remark this answers is not in the document]
Comment (olamy) [reply core/src/main/java/hudson/model/AbstractBuild.java:700]: it's a local variable.

File: core/src/main/java/hudson/tasks/LogRotator.java (hunk 13 above)
Comment (olamy) [reply core/src/main/java/hudson/tasks/LogRotator.java:220]: I do not see much difference neither if we have any gain and it's even the same length :)

File: core/src/main/java/hudson/util/Iterators.java (hunk 14 above)
Comment (olamy) [reply core/src/main/java/hudson/util/Iterators.java:321]: good catch

File: core/src/main/java/hudson/util/Iterators.java (hunk 11 above)
Comment (olamy) [reply core/src/main/java/hudson/util/Iterators.java:422]: so we can remove use of  `com.google.common.collect.Iterators` from RunList

File: core/src/main/java/hudson/util/RunList.java [hunk 16]
```diff
@@ -230,12 +221,8 @@ public boolean apply(@Nullable T r) {
      * @since 1.544
      * @deprecated use {@link #filter(Predicate)}
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/util/RunList.java:222]: remove as well then?

**Review by oleg-nenashev — COMMENTED** [review]
Looks good overall

File: core/src/main/java/hudson/util/RunList.java (hunk 16 above)
Comment (olamy) [reply core/src/main/java/hudson/util/RunList.java:222]: @oleg-nenashev I wish **REALLY** :)  but this comment https://github.com/jenkinsci/jenkins/pull/5344#discussion_r589122029

**Comment by basil** [comment]
The changes to `hudson.util.RunList` and `hudson.util.Iterators#removeNull` (which is consumed indirectly by `AbstractLazyLoadRunMap`) seem the most risky. The potential for regression here, like in `HistoryPageFilter`, is that we inadvertently turn a lazy-loading iterator into an eager-loading list. Loading runs from disk is a known performance bottleneck (especially on slow storage). Eagerly loading too many build records could also lead to memory exhaustion. Unfortunately `RunListTest` uses PowerMock and doesn't exercise the code in a realistic scenario. So I think this will need to be validated manually (if it hasn't been already).

**Comment by olamy** [comment]
@basil thanks for the review. I'm still struggle with LogRotator#perform which use HashMultimap and the usage is really strange :)

File: core/src/main/java/hudson/PluginWrapper.java [hunk 17]
```diff
@@ -220,7 +218,7 @@ public boolean hasDerivedDependencyErrors() {
      * The core can depend on a plugin if it is bundled. Sometimes it's the only thing that
      * depends on the plugin e.g. UI support library bundle plugin.
      */
-    private static Set<String> CORE_ONLY_DEPENDANT = ImmutableSet.copyOf(Collections.singletonList("jenkins-core"));
+    private static Set<String> CORE_ONLY_DEPENDANT = Collections.singleton("jenkins-core");
```
Comment (jtnord) [inline core/src/main/java/hudson/PluginWrapper.java:221]: 😆

**Comment by jglick** [comment]
> Unfortunately `RunListTest` uses PowerMock and doesn't exercise the code in a realistic scenario. So I think this will need to be validated manually

Use `RunLoadCounter` in JTH to verify lazy-loading issues. And yes be careful because mistakes will cause terrible performance issues in production that would not be noticed in casual interactive testing.

**Comment by olamy** [comment]
closing it as it looks to be done in few other PRs. such https://github.com/jenkinsci/jenkins/pull/5421
